_posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

```diff
@@ -0,0 +1,2334 @@
+  make sure the artifact is immutable (i.e., you never change it), and that it has a unique version number (so you can
+  distinguish this artifact from all the others).
+
+Infrastructure Modules::
+  Infrastructure modules are typically consumed as a library in the tool. Most infrastructure as code tools consume
+  libraries directly from a Git repository. For example, with Terraform you can consume modules through module blocks
+  that reference a Git repository (see
+  https://www.terraform.io/docs/configuration/modules.html[the official documentation] for more details). In this case,
+  using a Git tag to mark a revision with a human friendly name is sufficient to generate the release artifact.
+
+Live Infrastructure Config::
+  For live infrastructure config, there is typically no release artifact. Live infrastructure code doesn't need to be
+  packaged to deploy as it is directly consumable. For example, for Terraform or Terragrunt live config, you can
+  directly run `terraform apply` or `terragrunt apply` on the repo. In general, it is not necessary to tag your commits
+  for live infrastructure config because in practice you will end up deploying every commit of trunk.
+
+
+It is worth expanding a bit on the reason why live infrastructure config does not have any release artifact. To
+understand this, it is worth looking at what it means to have a working trunk on live infrastructure config. If you
+recall from <<run_the_code_locally>>, the only way to test live infrastructure config is by doing a dry run of the code.
+If the only way to test live infrastructure config is with dry runs, then you would want to make sure that there are no
+new changes to make to the live environments when you start. This is so that you get an accurate representation of the
+changes that are being introduced, since you don't want to be differentiating between existing changes that will be
+applied from trunk and the changes that will be applied with your new code.
+
+Given that, the definition of a "clean build" for the trunk with live infrastructure config is that a dry run returns no
+changes to make. This in turn means that the latest state of trunk that you are working off of should be a
+representation of what is actually deployed in your environments. Therefore, to ensure the trunk is clean, you will need
+to make sure that you continuously deploy and apply the trunk as new code is merged in.
+
+This leads to what we call _The Golden Rule of Infrastructure Code:_
+
+*_The master branch of the live repository should be a 1:1 representation of what's actually deployed._*
+
+You will want to do everything that is in your power to maintain this representation to streamline your development.
+
+
+[[deploy]]
+==== Deploy
+
+Now that you have a release artifact, the final stage of the process is to deploy the code. The act of deploying the
+code immediately after generating the release is Continuous Deployment (CD).
```

Comment:
Ah I didn't know that distinction. I am also not sure a lot of people know that distinction either. Do you think it is worth describing that in this document somewhere?

In any case, I am going to remove that sentence to avoid the accuracy problem and avoid definitions of terms, since it is really just fluff.

Reply:
It's from the book *Continuous Delivery**, IIRC. No need to define it here, just want to be sure we use the right terms.